### ISSUE TITLE
[$500] Document reproducible steps for "Characters missing when typing message in compose box"

### ISSUE THREAD

**Issue by mallenexpensify** [Expensify/App#34467]
Coming from [here in #expensify-open-source](https://expensify.slack.com/archives/C049HHMV9SM/p1705090464307259)

Occasionally and apparently randomly, when typing in the compose box on NewDot, characters will disappear or be removed. 

I've tried to document steps that can be reliably reproduced with no luck. If you're able to, post them and we'll compensate $250. 

If needed, comment in here or on the linked Slack thread above. 
<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01cd323eccdbb4d53d</li>
        <li>Upwork Job ID: 1745951741787557888</li>
        <li>Last Price Increase: 2024-02-29</li>
    </ul>
</details>

**Comment by mallenexpensify** [comment]
@getusha I don't think your help is needed her so I've unassigned.  If we need a C+ I'll add you back in.

**Comment by wildan-m** [comment]
@mallenexpensify which platform was the last time it occurred? can you forward the mentioned Slack post to #expensify-open-source ? or cite the important info here

**Comment by Ahmed-Abdella** [comment]
Hi, @mallenexpensify 

I've identified the steps to replicate a bug related to keystrokes in the composer, happens when attempting to type while the composer is not focused. In the initial video, I attempted to type the word 'fight,' but it omitted the 'i' letter. This issue is not limited to a particular letter; it occurred for me with the second letter as well. It appears to be a bug associated with typing when the composer is not in focus."

1. Lose focus on the composer
2. Try typing fast and notice
3. you need to type fast

https://github.com/Expensify/App/assets/77763699/6c667944-042a-473b-8bec-217cc64a5828

Another issue related to this and I think it has  the same root cause

steps to reproduce
1. Lose focus on the composer
2. write only one letter

result
the letter will appear twice.
and if you noticed immediately after typing the letter it appears twice for a moment then disappear and then it appears twice


https://github.com/Expensify/App/assets/77763699/1228070f-db23-4492-a5ba-37e3ceb17ac6

**Comment by wildan-m** [comment]
Here is my reproduction step. No need to type fast.

# One

1. Open report, and make sure the composer is focused and the text pointer blinking
2. Press Tab, the focus will be on the emoji icon
3. Type any character e.g. P

# Two
1. Open report, and make sure the composer is focused and the text pointer blinking
2. Press Tab twice, the focus will be on the arrow icon
3. Type any character e.g. P


https://github.com/Expensify/App/assets/11847279/c32e5cc1-ca89-4d9a-a3c0-7e2d7d722c50

**Comment by Ahmed-Abdella** [comment]
this is the second issue in my comment and I didn't say you have to type fast in this
you don't have to press tab you just lose the focus on the composer by doing this. you did the same as my second issue

**Comment by wildan-m** [comment]
@Ahmed-Abdella you didn't specify to which component the focus should be lost, I specified using the tab key and to emoji icon or arrow icon, I think that can be considered different

**Comment by Ahmed-Abdella** [comment]
Sorry @wildan-m but it is not specific to any component it just happened when you type while the compose is not focused. emoji icon and arrow icon have nothing to do here. and they are not different.

**Comment by bernhardoj** [comment]
FYI, the double character input is being handled [here](https://github.com/Expensify/App/issues/33710)

**Comment by mallenexpensify** [comment]
> which platform was the last time it occurred? 

I'm on Desktop, Mac app. I'm  unsure if it happens on web, I'll try to use that more to test. Just did two tests on web with no luck (which isn't surprising because it's really hard for me to reproduce on desktop) 

@Ahmed-Abdella , the issue you're running into is different. That's[ this GH issue](https://github.com/Expensify/App/issues/31286) coming from [this Slack thread](https://expensify.slack.com/archives/C049HHMV9SM/p1699896354633069)
That's a big pet peeve of mine, especially using chat switcher.  

@wildan-m the issue we're wanting repro steps here for is specially for characters not being recognized when typing in the compose box (and not just at the very beginning of a chat).  I haven't encountered the bug when using emoji or tabbing while in typing.

**Comment by Ahmed-Abdella** [comment]
> the issue you're running into is different. That's https://github.com/Expensify/App/issues/31286 coming from [this Slack thread](https://expensify.slack.com/archives/C049HHMV9SM/p1699896354633069)
That's a big pet peeve of mine, especially using chat switcher.

@mallenexpensify  The issue you mentioned here is not the same as mine, The keystrokes are actually recorded and typed, but there is a delay in the process.

My issue is that keystrokes are not being recorded or typed. ( characters missing ) like the issue you descriped

**Comment by mallenexpensify** [comment]
@Ahmed-Abdella can you elaborate on 'composer is not focused'?  

I just had it happen... twice! 
<img width="289" alt="image" src="https://github.com/Expensify/App/assets/22508304/87b47c20-9ee0-4e50-8708-dd11f63e81a7">
One hunch I have is that it happens more frequently after waking up your computer from sleeping, especially if the lid has been closed (I'm convinced there is a difference between putting your computer to sleep and closing the lid).

**Comment by Ahmed-Abdella** [comment]
@mallenexpensify just click outside the composer, and then start typing.
'composer is not focused' means that the composer doesn't have the green border, just click outside the composer, and start typing. like my first video above

This app feature shifts the focus to the composer when typing while the composer is not focused.

**Comment by wildan-m** [comment]
Agree that the lost focus issue might be the same with https://github.com/Expensify/App/issues/33710

**Comment by mallenexpensify** [comment]
Thanks @Ahmed-Abdella , I didn't know I could click outside of the compose box, start typing, then have text show up in the compose box.  When I tried that, I did notice that characters were missing, very often the second character (or... one of the first).  Do you think this is related to characters _randomly_ not being recognized after someone has been typing for a while (and not navigating outside of the compose box at all)?

@getusha bringing you back in for 👀 and help plz

**Comment by mallenexpensify** [comment]
Another thought/idea/consideration. 
I've def noticed that it sometimes takes a second for a keystroke to be recognized and appear.   I wonder if the bug where characters are missing is related to characters not immediately showing up then there's some sort of 'override' where, instead of adding the missing keystroke, NewDot skips it and shows the next letter you typed.

**Comment by getusha** [comment]
This issue is probably related to the composer being unfocused, yet the green border is displayed. I am trying to reproduce it.

**Comment by mallenexpensify** [comment]
What makes me think it's NOT related to the compose box being unfocused is that the bug often happens while I'm in the middle of typing a lot of text, when I'm focused (and actively typing) in the compose box. Here's an example that just happened.  I never left focus when typing this yet the `sp` at the end, where I typed `spot` don't show
<img width="807" alt="image" src="https://github.com/Expensify/App/assets/22508304/c8b056da-b132-408c-a5c2-962eefdb9e07">

**Comment by Ahmed-Abdella** [comment]
I am not sure, but It can be the same root cause.

**Comment by wildan-m** [comment]
@mallenexpensify, has anyone else experienced this issue? Have you tried using another keyboard? If you're using a wireless one, there might be a delay or a hardware issue with the keyboard.

**Comment by mallenexpensify** [comment]
Not sure if linked before but @tgolen is also experiencing.  [Slack chat here in #expensify-bugs](https://expensify.slack.com/archives/C049HHMV9SM/p1705090464307259)

I'm using the native keyboard on mac.  The issue is persistent on native and the same on wireless apple keyboard.  It's been happening for many weeks and it's happened 100+ times for me.   The weirdest part is that the timing of when it happens is totally random to me.  I can type 100 characters fine then one will get missed.  Other times, like mentioned above, it'll happen near the beginning of composing a chat.

**Comment by tgolen** [comment]
Yeah, I just had this happen again last week, totally out of the blue. Mine was in the middle of a two-sentence comment. I use staging/web with an external keyboard (that is hard-wired through a docking station). It seems to happen much more rare now.

If it was a hardware or keyboard issue, I would expect the problem to show up in any application, but it only happens to NewDot, so I do not think it is hardware related.

**Comment by tgolen** [comment]
OK, I just had it happen twice for me today, and I think I figured something out. I think it's related to the chat history loading. For example, this was the flow that causing a noticeable glitch for me:

1. Access a chat you haven't opened for a while (and therefore a lot of history has to be loaded)
2. While the chat is loading, begin typing your comment
3. Once the chat finishes loading, all I had typed so far just disappeared (about a full word or two)

I suspect this can happen on a smaller scale with chats you visit frequently where something having to do with the chat loading ends up destroying some of what you've typed.

**Comment by tgolen** [comment]
Note: I cannot reliably reproduce this, even if I throttle my network to 3G speed :(

**Comment by mallenexpensify** [comment]
I've also experienced it, at least, twice today.  One, I think, was at/after tapping the delete key. What I find interesting is it semi-frequently happens **after** I've typed a couple sentences, 100+ character.  After the initial load of a chat, are we periodically fetching or checking for more data?

**Comment by mallenexpensify** [comment]
I was able to capture the bug in action. In the vid, I typed `blurps` and the `b` was _removed_ or not recognized. 
https://github.com/Expensify/App/assets/22508304/c5998f49-d786-489a-9481-a2675e07f557

**Comment by mallenexpensify** [comment]
Only though, related to example above is to do another test and have the console open and see if can sync the time of the bug with something showing in the console.  Will try next week. 

IF anyone needs access to a high traffic workspace to test, drop your email in here and I'll add ya

**Comment by wildan-m** [comment]
@mallenexpensify count me in please, wildevemail+1@gmail.com is it different with high-traffic accounts?

**Comment by mallenexpensify** [comment]
Added @wildan-m !  Not sure when/how you get access, please confirm here once you have (it's a google form I fill out)

**Comment by wildan-m** [comment]
Is it the WS? seems like a typical workspace, what are the differences?

<img width="1436" alt="image" src="https://github.com/Expensify/App/assets/11847279/43a8079f-adbd-4f61-a1c8-1db3e7283581">

**Comment by mallenexpensify** [comment]
> @mallenexpensify Is it the WS? seems like a typical workspace, what are the differences?

@wildan-m , yes it is.  I think the difference is that there are a ton of members and chats on the WS, even though it's not visible to you (Which I don't quite understand, I only see a few members in there and you're not one of them.  Do you see mallen@expensifail.com listed anywhere? )
<img width="407" alt="image" src="https://github.com/Expensify/App/assets/22508304/c6632e1e-aff5-4a06-8c18-410edfb0fde6">

**Comment by tgolen** [comment]
I think the problem is getting slightly worse. Today I was typing the word "that". I saw it erase the first `t` and then place it behind the `h` so I ended up with `htat`. I know that I'm normally not the best speller, but I do know the difference between me making a typo mistake, and something funky happening in a text input that changes my input into a typo.

This is slightly different than what I was previously seeing where the `t` would have simply been gone.

**Comment by tgolen** [comment]
Matt, do you use Grammerly at all? I do, and I had a thought that this
could possibly be the culprit. I'm going to disable it and see if I notice
any difference.

**Comment by mallenexpensify** [comment]
I don't use Grammarly, is yours software on your computer or a chrome extension? 

Tim was able to reproduce while testing something out, see the `?` disappear at the end of the vid below. 

https://github.com/Expensify/App/assets/22508304/2425e563-cebc-4578-9681-e26853d9dd57

**Comment by mallenexpensify** [comment]
Had it happen at least twice today.  Tried to capture in vid with logs but....not luck.  Bumping back to External and help wanted.

**Comment by jeremy-croff** [comment]
I'm able to consistently reproduce. With the help of Emoji's as they require more parsing and slow down the chat more.

https://github.com/Expensify/App/assets/157416545/0b894855-1c50-4740-9ab3-357797967e12

Documented steps:
Go to Composer
type 123, select emoji type 123 select emoji -> send 
repeat

Technically this is possible on any value text input with a controlled value prop, depending on lag

**Comment by Ahmed-Abdella** [comment]
> @mallenexpensify I'm able to consistently reproduce. With the help of Emoji's as they require more parsing and slow down the chat more.
> 
>  cursor.issue.mov 
> Documented steps: Go to Composer type 123, select emoji type 123 select emoji -> send repeat
> 
> What makes me think this is the use case, is that it also shows the character (emoji) before deleting it by itself. Technically this is possible on any value text input with a controlled value prop, depending on lag
> 
> Also! the pattern is so weird. check the last emojis. Miss Miss, Emoji Emoji, Miss Miss, Emoji Emoji Where misses are when it deletes it


"This is a follow-up to the focus issue I previously discussed. When you click on the Emoji picker to open the emojis, it causes you to lose focus on the composer. As a result, when you start typing the emoji, you may miss the character or emoji you intended to type since the composer is not focused."

**Comment by Ahmed-Abdella** [comment]
> I think the problem is getting slightly worse. Today I was typing the word "that". I saw it erase the first `t` and then place it behind the `h` so I ended up with `htat`. I know that I'm normally not the best speller, but I do know the difference between me making a typo mistake, and something funky happening in a text input that changes my input into a typo.
> 
> This is slightly different than what I was previously seeing where the `t` would have simply been gone.

@tgolen I can confirm this, Happened alot  for me

**Comment by jeremy-croff** [comment]
> > @mallenexpensify I'm able to consistently reproduce. With the help of Emoji's as they require more parsing and slow down the chat more.
> 
> > 
> 
> >  cursor.issue.mov 
> 
> > Documented steps: Go to Composer type 123, select emoji type 123 select emoji -> send repeat
> 
> > 
> 
> > What makes me think this is the use case, is that it also shows the character (emoji) before deleting it by itself. Technically this is possible on any value text input with a controlled value prop, depending on lag
> 
> > 
> 
> > Also! the pattern is so weird. check the last emojis. Miss Miss, Emoji Emoji, Miss Miss, Emoji Emoji Where misses are when it deletes it
> 
> 
> 
> 
> 
> "This is a follow-up to the focus issue I previously discussed. When you click on the Emoji picker to open the emojis, it causes you to lose focus on the composer. As a result, when you start typing the emoji, you may miss the character or emoji you intended to type since the composer is not focused."

What makes you consider this a focus issue? The composer seems focused throughout the bug. Additionally it is rendering the emoji in the composer, and then deleting it. If it was unfocused it wouldn't render the emoji in the composer to begin with.

**Comment by Ahmed-Abdella** [comment]
After researching and tests, It seems that this issue is not related to our application. Rather, it seems to be a problem with the MacOS system. I encountered this issue while using Slack, and upon browsing through various Apple discussions, I noticed that many others have faced similar issues.

It is happening regularly while typing fast

![Screenshot 2024-02-12 at 4 23 50 AM](https://github.com/Expensify/App/assets/77763699/12cee591-a5e7-4fa7-8290-6578ec591935)

![Screenshot 2024-02-12 at 4 08 41 AM](https://github.com/Expensify/App/assets/77763699/80fbe1d0-b298-4e25-b4cd-9546e4527d19)

Another weird behavior that supports this issue is that I tested typing a letter followed by a space and then another letter, and repeated this process. When I typed slowly, nothing unusual happened. However, when I typed quickly, a random "." was added after some letters. This behavior occurred not only in our application, but also in Github and Slack, and looks like it happens everywhere

Expensify 

https://github.com/Expensify/App/assets/77763699/a880f757-5545-48f9-8891-9dafa002e8e6

Github

https://github.com/Expensify/App/assets/77763699/aacdd661-0378-4927-bda5-800c551de456

So, yeah in the end it looks like that the characters missing issue is not related to Expensify, it is about MacOS.

@mallenexpensify @tgolen  
You can test this through other applications to confirm this. Also search google for "characters missing on MacOS" or "letters missing on MacOS" and you will find alot of forums and discussions about this behavior, I can provide you some links if you want.

**Comment by Ahmed-Abdella** [comment]
@jeremy-croff in your state When you click the icon that opens the emoji picker the composer lose focus.

And when you choose the emoji it wil be renderd in the composer.
It renders the emoji while the composer is not focused, yeah this actually happens. this behavior is a feature, when you start typing (normal typing, or adding emojis) while the composer is not focused the composer will be focused and the letters added to it.

(Try it by clicking outside the composer and then start typing)

**Comment by jeremy-croff** [comment]
> Only though, related to example above is to do another test and have the console open and see if can sync the time of the bug with something showing in the console.  Will try next week. 
> 
> 
> 
> IF anyone needs access to a high traffic workspace to test, drop your email in here and I'll add ya

Jeremycroff@gmail.com

**Comment by dragnoir** [comment]
> IF anyone needs access to a high traffic workspace to test, drop your email in here and I'll add ya

mosaixel.org@gmail.com

**Comment by dragnoir** [comment]
## steps to reproduce the issue.

### 1: Open the app with chrome browser where you have devtools installed

### 2: Disable cache and decrease CPU performance to 6x slowdown

![image](https://github.com/Expensify/App/assets/12425932/87c6a3cb-3da0-4c79-93ea-9264323cde37)

### 3: Start typing on old conversations with lots of chat
If you already tested on this chat, logout and login or clean the local Onyx database.
The moment the chat is loaded, there will be an overwelmed tasks on the CPU wish will cause the delay on the input speed

![image](https://github.com/Expensify/App/assets/12425932/9a338cb0-2a63-48f1-9a94-e44d43b6182a)

**Comment by jeremy-croff** [comment]
If you slowdown the CPU, you can reproduce the emoji showing and deleting everytime:

https://github.com/Expensify/App/assets/157416545/a61bb08e-c67f-4a12-b356-d8dfb8cd0f87

It's also CPU related not purely focus related, as with less lag in the previous experiment it happened every 2 times

**Comment by mallenexpensify** [comment]
@jeremy-croff , @dragnoir , I've added you to high traffic workspace so you can test there. 

On MacOS, I also don't experience the issue in/with other applications.  I've noticed it 100+ times in the Desktop app and I would have  noticed in Slack, gmail and other spots if it happened frequently.  

Thanks everyone for chiming in here, let's keep this discussion going til we are able to definitely reproduce.

**Comment by mallenexpensify** [comment]
Here's a weird twist... 
I was editing a block of text, midway through, the cursor went back one space then relocated to the end of the text block.   So.. maybe we most-often see the bug as just affecting the last character or two but it really affects one character then the cursor refocuses to the end.

**Comment by tgolen** [comment]
I have witnessed that as well, Matt. It seems to happen less frequently,
but I feel it's related.

Also, to report back about Grammarly, disabling it hasn't stopped this from
occurring.

**Comment by wildan-m** [comment]
Hi, I have identified the bug.

The bug can be triggered by pressing a keyboard character every second.

Ideally, the interval should be around 1 second to consistently reproduce it.

While it is possible to do it manually, it can be cumbersome as hitting the keyboard at exact intervals is challenging.

It is more efficient to automate this process using a macro application, such as AppleScript in my case.

```applescript
on run
	repeat
		tell application "System Events"
			keystroke "W"
			delay 1
		end tell
	end repeat
end run
```
Here is some information about the issue:
- I encounter it more frequently with wider text characters, in my test 'W' letter triggers more bugs than 'c'
- The issue can be replicated in any chat type: Workspace / 1:1 / Room
- If it occurs less frequently after a long text, double-clicking the text will reset the existing text and it usually happens after around the 6th stroke
- I encounter it more frequently with the Chrome browser than the desktop app (both of which are actually Chrome)

Here is the video to replicate it (Left: Script, Middle: Chrome Browser, Right: Desktop App)


https://github.com/Expensify/App/assets/11847279/c5a0a9fb-0713-48bc-ae28-1637e35f9a8f

**Comment by Victor-Nyagudi** [comment]
**Version Number:** Latest (going off of @mallenexpensify being able to reproduce the bug recently)
**Reproducible in staging?:** y (based on Matt reproducing it)
**Reproducible in production?:** y (I'm assuming Matt uses the production desktop app)
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** 
**Expensify/Expensify Issue URL:**
**Issue reported by:** @Victor-Nyagudi 
**Slack conversation:** https://expensify.slack.com/archives/C01GTK53T8Q/p1704311664219129

## Action Performed: 

Pre-condition: You'll need to be able to see the Onyx merge calls, so have the console open. 

1. Open a chat with very many messages/long chat history. 
2. Type a few letters in the composer so that the most recent Onyx call is a `merge()`. You should see a message like this:

```
[info] [Onyx] merge() called for key: reportDraftComment_31984933430 (this is a random number I just came up with) 
```

3. **This is the difficult part.** As you type a word, try and match your keystrokes such that you synchronize the character you type with the next merge call log on the console. 

In other words, type a couple of letters just as the number to the left of the merge call increases. You'll need to try this multiple times, but if you do it just right, some letters will be deleted and maybe retyped.

I've had some slightly better luck when trying to reproduce this offline. 

Additionally, I was able to reproduce this by after clicking on the composer, moving the cursor away from it and then trying to synchronize keystrokes with the Onyx merge calls. 

## Expected Result: 

The user is able to type characters and words as usual without any problems. 

## Actual Result:

Some characters are randomly deleted. 

## Workaround: 

It's rare for a user to time their keystroke perfectly with the Onyx merge call, so even though it happens, it does so rarely, and a user can proceed as usual. 

## Plarforms: 

Which of our officially supported platforms is this issue occurring on?

- [ ] Android Native
- [ ] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [x] MacOS: Desktop

## Screenshots/Videos 

Add any screenshot/video evidence

<details>

<summary>Mac Desktop</summary>

https://github.com/Expensify/App/assets/79470910/ae1fb3cf-5655-4dc2-a6dd-b1e7e7ca7426

Above, I wanted to write "just as I'm typing" but the "I" in "I'm" got deleted. 

https://github.com/Expensify/App/assets/79470910/7cb23f1a-c370-4d4d-bef1-0fb5fe64ab62

I wanted to type "maybe another" but the "a" got deleted there at the end. 

https://github.com/Expensify/App/assets/79470910/27902c9c-19be-4620-b57d-17ec8774ea13

The "t" in "to" got deleted. Did this one while offline. 

</details>

<details>

<summary>Mac Chrome</summary>

https://github.com/Expensify/App/assets/79470910/d3d2bd23-41ac-494b-b6a7-70477873fa95

The "w" in "with" got deleted here. 

</details>

<details>

<summary>Mac Safari</summary>

https://github.com/Expensify/App/assets/79470910/f31711d0-484c-4da4-b30a-13f88515323c

I wanted to type "maybe I'm" but the "I" and "m" got deleted. 

</details>

If you keep your eye on the number beside the most recent Onyx merge call in the console in all the videos the moment the character is erroneously deleted, you'll notice that two calls happen in rapid succession.

**Comment by jeremy-croff** [comment]
> Hi, I have identified the bug.
> 
> The bug can be triggered by pressing a keyboard character every second.
> 
> Ideally, the interval should be around 1 second to consistently reproduce it.
> 
> While it is possible to do it manually, it can be cumbersome as hitting the keyboard at exact intervals is challenging.
> 
> It is more efficient to automate this process using a macro application, such as AppleScript in my case.
> 
> ```applescript
> on run
> 	repeat
> 		tell application "System Events"
> 			keystroke "W"
> 			delay 1
> 		end tell
> 	end repeat
> end run
> ```
> 
> Here is some information about the issue:
> 
> * I encounter it more frequently with wider text characters, in my test 'W' letter triggers more bugs than 'c'
> * The issue can be replicated in any chat type: Workspace / 1:1 / Room
> * When using the above script, if it occurs less frequently after a long text, double-clicking the text will reset the existing text and it usually happens after around the 6th stroke
> * I encounter it more frequently with the Chrome browser than the desktop app (both of which are actually Chrome)
> 
> Here is the video to replicate it (Left: Script, Middle: Chrome Browser, Right: Desktop App)
> 
>  GMT20240216-101429_Clip_Wildan.M.s.Clip.02_16_2024.mp4

Does this mean OP is a slow typer?

**Comment by Victor-Nyagudi** [comment]
**TLDR of what I suspect is going on:**

1. User tries typing a word e.g. "chocolate".
2. The keystroke for the "c" at the beginning happens simultaneously with an Onyx merge call. This can happen with any character in the word based on timing. 
3. Composer shows "chocolate" briefly but is re-rendered with what's in Onyx - "hocolate".

**Long answer**

While doing my investigation, I noticed we're dealing with a `valueBeforeCaret` variable which is updated when `shouldCalculateCaretPosition` is `true`. 

https://github.com/Expensify/App/blob/28a1ccc40ad6345338de1951c02146304fb30cd5/src/components/Composer/index.tsx#L126-L131

Since `setValueBeforeCaret` is fired here, this should trigger a re-render. 

However, before the re-render occurs, the value of `valueBeforeCaret` at this point will be everything you've typed minus the most recent character e.g. "I love chocolat"

I'd like to believe that before this re-render happens, if you time your keystroke perfectly, the letter you just typed (the last "e" at the end of the word "chocolate") shows up just before the composer re-renders.

However, this happens at the same time an Onyx merge is called, thus, for a brief moment, you see the most recent character you typed before the merge completes resulting in what's stored in Onyx ("I love chocolat" - the draft comment) at this point replacing what's currently in the composer ("I love chocolate") as the component re-renders because a `setState` got called. 

As for the 2 merge calls happening in quick succession, I suspect the first is the initial one that should've happened when you stopped typing briefly, and the second is for when you resumed.

**Comment by jeremy-croff** [comment]
I can confirm reproduction manually, using a metronome (lol). Highly increased occurrences when you sow down CPU speed.

I have stacked all the findings:
1. Grammarly 
2. Electron
3. Slow CPU speed
4. paced typing at 1 char per sec
5.  Console open for info

I added in alternative characters and did it manually to show it can happend to a normal person:
I am Typing WUWUWUWU here:

https://drive.google.com/file/d/1xDxS-zwgItYBKB5ytz7W2TShgspB6Ioq/view?usp=drive_link

**Comment by jeremy-croff** [comment]
It  seems in our composer we have debouncedSaveReportComment, at 1000ms. When lowing to 500ms for example, we cannot reproduce the bug. I only seem to be able to reproduce it when I sync the time speed to the debouncedSaveReportComment setInterval. And increase occurence with CPU slowdown ( maybe bigger race condition zone )

**Comment by jeremy-croff** [comment]
This code was added 5 months ago and could be the culprit of all this mess

**Comment by wildan-m** [comment]
@mallenexpensify @getusha @tgolen Can you replicate it following [the steps I posted](https://github.com/Expensify/App/issues/34467#issuecomment-1948125511)?


> Does this mean OP is a slow typer?

@jeremy-croff No, but sometimes we type at a 1-second interval and encounter the bug. I suspect it's due to the 1000ms `debouncedSaveReportComment`. There could be a race condition causing the bug, but further analysis is needed to assess and determine the optimal solution. Since this post is meant to document the reproducible steps, I think we need to create a new post to discover the optimal solutions.

**Comment by tgolen** [comment]
Wow, sounds like we are finally on the right track! Keep up the good work!

**Comment by wildan-m** [comment]
@mallenexpensify @getusha I've updated [my submission](https://github.com/Expensify/App/issues/34467#issuecomment-1948125511).

It include [@mallenexpensify's case](https://github.com/Expensify/App/issues/34467#issuecomment-1920050760) when typing 'blurps'

**Comment by mallenexpensify** [comment]
Thanks again @wildan-m @Victor-Nyagudi and @jeremy-croff for the continued testing and feedback.  
@getusha , 👀 please, as you're the C+.  Please review and test the above to see which of the above is the most likely culprit.  @jeremy-croff provided a handy list [here](https://github.com/Expensify/App/issues/34467#issuecomment-1948560119)

> Does this mean OP is a slow typer?

I could see it maybe being related to a specific cadence, I feel like it _might_ often happening after I pause for a moment when typing ¯\_(ツ)_/¯.

**Comment by getusha** [comment]
>Grammarly

looks like @tgolen was able to repro after disabling it😬

> Electron

An electron bug? @jeremy-croff, if yes, i already tried to find if the issue has been logged in their repo. 

> Slow CPU speed

Just tried this multiple times, no luck.

**Comment by getusha** [comment]
@mallenexpensify Was able to reproduce this only once using @wildan-m's script. trying it again.

https://github.com/Expensify/App/assets/75031127/405eb8bd-c4d1-405b-985d-c26704a8c720

**Comment by wildan-m** [comment]
@getusha, please add my second script too. Sometimes it slips our eyes, but incomplete blurps word clearly indicate missed keystrokes.

**Comment by jeremy-croff** [comment]
> @mallenexpensify Was able to reproduce this only once using @wildan-m's script. trying it again.
> 
> 
> 
> 
> 
> https://github.com/Expensify/App/assets/75031127/db1e5ed9-f0e9-445a-92bc-99b948cb0baf
> 
> 
> 
> 

Did you try the metronome and slowed CPU speed? This is the only way to manually reproduce it.
Maybe I've should've been more clearer.
It requires a lot of focus manually to time it right.

I also believe I proofed out that it's related to the code I referenced, because after editing those values even the script cannot reproduce it anymore.

**Comment by getusha** [comment]
@jeremy-croff was able to reproduce it using `metronome` without slowing down the cpu

https://github.com/Expensify/App/assets/75031127/d3072cf8-b523-4d95-b23a-19cc3b8d2cac

the question is how reliable is this? for example, the script worked once then i am not able to repro again. if i can reproduce this at least 6 times out of 10. we can proceed with it.

**Comment by jeremy-croff** [comment]
> @jeremy-croff was able to reproduce it using `metronome` without slowing down the cpu
> 
> 
> 
> https://github.com/Expensify/App/assets/75031127/d3072cf8-b523-4d95-b23a-19cc3b8d2cac
> 
> 
> 
> the question is how reliable is this? for example, the script worked once then i am not able to repro again. if i can reproduce this at least 6 times out of 10. we can proceed with it.
> 
> 

I seem to be able to reproduce it also without slowing down CPU, just much harder.

And I can consistently reproduce it with the script by sending the message after it occurs once, as it seemed less likely to happend twice in the same message

**Comment by mallenexpensify** [comment]
It's particularly bad for me right now in chat with Chronos for drafting my 10am. Had, at least, 3 instances where characters were _missed_.  

@getusha (and others, if you want) can you review the vid below to see what shows in the console when the `l` is removed at the end of the test?  I'm wondering if that has anything to do with the bug.  

https://github.com/Expensify/App/assets/22508304/6131744a-532a-4b6a-8188-249789ff51a2

**Comment by mallenexpensify** [comment]
Just happened again in a different chat and it looks to be the same in the console at that time! 

<img width="384" alt="image" src="https://github.com/Expensify/App/assets/22508304/9183528f-099d-4e2e-bb05-3391eef58b90">

**Comment by getusha** [comment]
Interesting, was able to track that log to this line,
https://github.com/getusha/App/commit/98bb17747372770ac08fe6847fcee0ac826383b5#diff-3eab019f4e3637c02ef9e920269b24b28d08b76356021c19c4d117ef63c54277R191-R193
as @jeremy-croff mentioned.
i think yes it is related to that,

**Comment by jeremy-croff** [comment]
Yeah, I thought posting manual reproduction steps, as well as a single line of code you can change to mitigate the bug was a slam dunk for an RCA :P

**Comment by wildan-m** [comment]
> @getusha (and others, if you want) can you review the vid below to see what shows in the console when the l is removed at the end of the test? I'm wondering if that has anything to do with the bug.

@mallenexpensify Yes, the onyx save trigger happens when we call `saveReportComment`, and it will happen more frequently when we type at a one-second interval (our current debounce interval).

**Comment by wildan-m** [comment]
> the question is how reliable is this? for example, the script worked once then i am not able to repro again. if i can reproduce this at least 6 times out of 10. we can proceed with it.

@getusha Utilize the script to select words that have minimal intersections and a broader range of characters.

WA -> It has intersection
Hello -> contain double "l" that have less character width

I can produce more frequently using the words "blurps" and "JUMP"
5 out of 16
![image](https://github.com/Expensify/App/assets/11847279/a19a692e-105e-4372-ba74-e605b9988ee0)

**Comment by wildan-m** [comment]
@getusha @mallenexpensify To increase error occurrences, we can reduce the debounce interval to 100 ms.

Change the code to:

```
    const debouncedSaveReportComment = useMemo(
        () =>
            _.debounce((selectedReportID, newComment) => {
                Report.saveReportComment(selectedReportID, newComment || '');
            }, 100),
        [],
    );
```

And change test script to:

```applescript
on run
	set textToType to "hello "
	repeat
		emulateTyping(textToType)
		delay 1
	end repeat
end run

on emulateTyping(textToType)
	tell application "System Events"
		repeat with i from 1 to count of characters of textToType
			set currentChar to character i of textToType
			keystroke currentChar
			delay 0.1
		end repeat
	end tell
end emulateTyping
```

It's important to keep a 1-second delay between each word loop; otherwise, we won't be able to manually escape the script loop and will need to force restart macOS.

https://github.com/Expensify/App/assets/11847279/ec8aca98-9e7c-4384-b95a-fbc4f1bffd71

<img width="311" alt="image" src="https://github.com/Expensify/App/assets/11847279/3b177bb1-796d-4354-ad61-fbdc6211d767">

**Comment by AndrewGable** [comment]
I have had this happen to me when I am typing really fast

**Comment by wildan-m** [comment]
> I have had this happen to me when I am typing really fast

@AndrewGable, troubleshooting is challenging without recording it. My assumption is that you take breaks while typing, triggering the debounce interval. I suspect a potential race condition between the debounce interval and our custom typing handle, causing it to not consistently occur.

**Comment by mallenexpensify** [comment]
> My assumption is that you take breaks while typing, triggering the debounce interval. I suspect a potential race condition between the debounce interval and our custom typing handle, causing it to not consistently occur.

This makes sense for my use cases.  The bug seems to happen semi-frequently but more frequently when I'm *not* typing at a consistent speed or if I pause, backspace/delete or slow/fasten my pace of typing.

**Comment by mallenexpensify** [comment]
The bug is much, much worse for me the past two days, anyone else?  I'm on Version 1.4.43-7, desktop. 
@getusha do we have reliable reproducible steps so that we can create an issue for the job?  If so, can you document them.  

If/when we create a job, y'all on here will have dibs for submitting proposals before it gets open to external contributors.  (we'll also discuss potential compensation for this issue once the job is created).

**Comment by getusha** [comment]
1. Open Script Editor (Mac)
2. New Document > Paste the following script
```bash
on run
	set textToType to "JUMP "
	repeat
		emulateTyping(textToType)
		delay 1
	end repeat
end run

on emulateTyping(textToType)
	tell application "System Events"
		repeat with i from 1 to count of characters of textToType
			set currentChar to character i of textToType
			keystroke currentChar
			delay 0.1
		end repeat
	end tell
end emulateTyping
```
3. Press the play button (run the script)
4. Focus the desktop newDot app composer
5. Wait until you observe that a letter from the word 'JUMP' is being omitted after being typed.

**Expected result:**
The word 'JUMP' needs to be typed with consistency in the composer. **ignore the first word it is because we had a delay when focusing the input.**

**Actual result:**
After a few tries character gets omitted and the word becomes inconsistent. i.e. `JUMP JUMP JMP`

https://github.com/Expensify/App/assets/75031127/65090f16-9610-456c-8490-201064151216

**Comment by getusha** [comment]
@mallenexpensify i was able to reproduce this 5/5 times running the script, try it and let me know.

**Comment by wildan-m** [comment]
@getusha from the video seems you mean `delay 1` instead of `delay 0.1`

**Comment by getusha** [comment]
oh thanks, i didn't notice that.

**Comment by mallenexpensify** [comment]
Me again!!  Desktop is the worst it's ever been so I closed it and was testing on Web/Chrome.  I was *unable* to reproduce on web, which makes me wonder if part of the issue is related to cache/cookies/storage or similar (not an engineer!). 

Based on the above, I sent into our new Troubleshooting feature, I cleared cache and restarted and it's *much* better. I wasn't able to reproduce the missing characters...  yet.  Will report back if/when I'm able to.  

<img width="448" alt="image" src="https://github.com/Expensify/App/assets/22508304/f80abfad-7b35-4560-aaae-96bb756998cf">

**Comment by mallenexpensify** [comment]
**Headsup** if you decide to `Clear cache and restart`, you need to sign out/in or else the app will be bricked (reported in #expensify-bugs)

**Comment by wildan-m** [comment]
@mallenexpensify have you tried reproducing it using a script? It can be difficult to manually hit the debounce interval accurately. I suggest adapting the script to your desired text and then using online tools like [text-compare](https://text-compare.com/) to compare the original and actual text. For example:

```applescript
on run
	set textToType to "   In a world where gravity seems to defy its own rules, where the laws of physics bend and twist, our daring adventurer embarks on a journey filled with uncertainty. Armed with courage and a relentless spirit, the fearless explorer takes a leap of faith, uttering the word 'JUMP' with each bound. The echoing sound of determination reverberates through the vast expanse, as the quest unfolds into an epic tale of leaps, twists, and unforeseen challenges. Brace yourself for an adventure like no other, where each keystroke marks a moment in the extraordinary saga of boundless exploration."
	emulateTyping(textToType)
end run

on emulateTyping(textToType)
	tell application "System Events"
		repeat with i from 1 to count of characters of textToType
			set currentChar to character i of textToType
			keystroke currentChar
			delay 1
		end repeat
	end tell
end emulateTyping
 
```

**Comment by mallenexpensify** [comment]
I have not @wildan-m , @getusha can you give it a shot?

**Comment by wildan-m** [comment]
@mallenexpensify I believe @getusha has validated the reproduction steps [here](https://github.com/Expensify/App/issues/34467#issuecomment-1958976159). Right?

**Comment by getusha** [comment]
@mallenexpensify was able to repro after clear cache & restart, are you still unable to reproduce it?

**Comment by mallenexpensify** [comment]
@wildan-m , thanks, it does look like [those](https://github.com/Expensify/App/issues/34467#issuecomment-1958976159) are the repro steps.  @tgolen , can you take a look since you're well-versed in the bug (and I'm not an engineer?

Ideally, I'm thinking:
1. Confirm we have reliable reproduction steps
2. Create a new issue with the steps, **don't** add the `Help Wanted` label, cross post here asking y'all to submit proposals in 48 hours before opening it up to others
3. Discuss compensation for this issue. 
4. Pay then Close this issue
5. Bug gets fixed in other issue 🎉

**Comment by mallenexpensify** [comment]
@rinej @allroundexperts (and others)
Do you think this issue is related to this one you worked on?
- https://github.com/Expensify/App/issues/28562

**Comment by tgolen** [comment]
Yes, I think the reproduction has been identified really well. I like your plan of action to move forward.

**Comment by mallenexpensify** [comment]
New issue created if you'd like to submit a proposal, price is $500 
- https://github.com/Expensify/App/issues/37356

I propose we pay @getusha $500 for C+
@getusha , can you propose what you think is fair payment for the contributor/s here? Thx.

**Comment by getusha** [comment]
@wildan-m - **75%**, first provided working reproduction steps, which we then followed.
@jeremy-croff - **25%** provided working alternative reproduction steps https://github.com/Expensify/App/issues/34467#issuecomment-1954795985, which can be helpful latter.

@mallenexpensify i think this is fair

**Comment by mallenexpensify** [comment]
Thanks @getusha , I like the breakdown and the two ❤️ from the contributors make me think they're 👍 with it too.  
@wildan-m , @jeremy-croff can you please apply here and I'll hire you? 
https://www.upwork.com/jobs/~01cd323eccdbb4d53d

Personally.. I'd like to thank everyone here. This issue has been hella frustrating for me for many weeks.  Also, I'm planning to implement a process for how to create 'document reproducible steps' issues in the future, since we have quite a few.

**Comment by wildan-m** [comment]
@mallenexpensify applied. Thank you!

**Comment by mallenexpensify** [comment]
Thanks @wildan-m and @jeremy-croff , can you please accept the job and reply here once you have? 
[upwork.com/jobs/~01cd323eccdbb4d53d](https://www.upwork.com/jobs/~01cd323eccdbb4d53d)

**Comment by jeremy-croff** [comment]
> Thanks @wildan-m and @jeremy-croff , can you please accept the job and reply here once you have? 
> 
> [upwork.com/jobs/~01cd323eccdbb4d53d](https://www.upwork.com/jobs/~01cd323eccdbb4d53d) 

I also replied yesterday! Thanks

**Comment by wildan-m** [comment]
@mallenexpensify accepted the offer. Thanks

**Comment by mallenexpensify** [comment]
Contributor: @wildan-m paid $375 via Upwork
Contributor: @jeremy-croff paid $125 via Upwork
Contributor+: @getusha owed $500 via Upwork. 

@getusha can you please accept the job and reply here once you have? 
https://www.upwork.com/jobs/~01cd323eccdbb4d53d

**Comment by getusha** [comment]
>@getusha can you please accept the job and reply here once you have?
@mallenexpensify accepted

**Comment by mallenexpensify** [comment]
@getusha paid, payment breakdown above is updated. 

Thanks everyone!!! I particularly hate this bug, and it just happened to me two mins ago.
